Disambiguate GenAPI calls to base constructors

ericstj opened the pull request:
Fix https://github.com/dotnet/sdk/issues/31655

ericstj committed:
Disambiguate GenAPI calls to base constructors

ViktorHofer reviewed (approved):
Lovely. When we wrote this, we believed that a default without specifying the underlying type is sufficient. We were wrong.

carlossanlop commented on src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:
```diff
@@ -1284,6 +1284,43 @@ public partial class C : B
+                    }
+                    """);
+        }
+
```
@ericstj do you mind also adding test cases for the examples provided in the issue? https://github.com/dotnet/sdk/issues/31655

One default is casted to a class type, the other one to a generic type:

```cs
base((Packaging.Core.PackageIdentity)default!, (System.Collections.Generic.IEnumerable<Packaging.Core.PackageDependency>)default!)
```

Want to make sure both work as expected.

ericstj replied:
Sure I can, though I don't think we need to use those NuGet types.  I'll add some user types.

ericstj replied:
Interesting, so I'm adding the type arguments to default correctly, but it's being removed by https://github.com/dotnet/sdk/blob/4ea3e8304c21022bbea72a29154b7602af4f9239/src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs#L94

This seems to be a bug in the Reducer in roslyn.

ericstj replied:
Filed https://github.com/dotnet/roslyn/issues/67867 and I will add the regression test with a suppression.

ericstj committed:
Add a test with ambiguous value type arguments for base constructor

This test currently fails because https://github.com/dotnet/roslyn/issues/67867
Simplifier.ReduceAsync overly simplifies the calls to base constructor
by removing type parameters.
